File: scripts/ci_provisioning.rb
require 'spaceship'
require 'base64'
require 'fileutils'

puts "🔐 Initializing Spaceship App Store Connect API Token..."
key_id = ENV['APP_STORE_CONNECT_KEY_ID']
issuer_id = ENV['APP_STORE_CONNECT_ISSUER_ID']
key_content = ENV['APP_STORE_CONNECT_API_KEY']
dev_team = ENV['DEV_TEAM'] || 'V7J345DY58'

key_path = File.expand_path("~/.appstoreconnect/private_keys/AuthKey_#{key_id}.p8")
if File.exist?(key_path)
  key_content = File.read(key_path)
end

token = Spaceship::ConnectAPI::Token.create(
  key_id: key_id,
  issuer_id: issuer_id,
  filepath: key_path,
  key: key_content,
  in_house: false
)
Spaceship::ConnectAPI.token = token

# 1. Setup Keychain and Intermediate Certificates
keychain_path = "/tmp/build.keychain"
password = "buildpassword"
puts "🔑 Configuring build keychain at #{keychain_path}..."
system("security create-keychain -p #{password} #{keychain_path} 2>/dev/null || true")
system("security set-keychain-settings -lut 21600 #{keychain_path}")
system("security unlock-keychain -p #{password} #{keychain_path}")

out = `security list-keychains -d user`.to_s
existing = out.split("\n").map { |k| k.strip.gsub('"', '') }.reject(&:empty?)
all_kcs = ([keychain_path] + existing).uniq
system("security list-keychains -d user -s #{all_kcs.join(' ')}")

# Download and install Apple Root CA and WWDR intermediate certificates
wwdr_urls = [
  "https://www.apple.com/certificateauthority/AppleWWDRCAG3.cer",
  "https://www.apple.com/certificateauthority/AppleWWDRCAG6.cer",
  "https://www.apple.com/appleca/AppleIncRootCertificate.cer"
]
wwdr_urls.each do |url|
  fname = File.basename(url)
  system("curl -s -f -L '#{url}' -o '/tmp/#{fname}' && security import '/tmp/#{fname}' -k #{keychain_path} -T /usr/bin/codesign -T /usr/bin/security 2>/dev/null || true")
end

# 2. Check / Create Distribution Certificate in Keychain & Developer Portal
active_cert = nil
dist_key_path = "/tmp/dist.key"
csr_path = "/tmp/dist.csr"
cer_path = "/tmp/dist.cer"
pem_path = "/tmp/dist.pem"
p12_path = "/tmp/dist.p12"

puts "🚀 Generating fresh private key and CSR for Apple Distribution..."
system("openssl genrsa -out #{dist_key_path} 2048")
system("openssl req -new -key #{dist_key_path} -out #{csr_path} -subj '/CN=Apple Distribution: Forma/O=#{dev_team}'")
csr_content = File.read(csr_path)

puts "📋 Checking distribution certificates on App Store Connect..."
dist_certs = Spaceship::ConnectAPI::Certificate.all(filter: { certificateType: "DISTRIBUTION" })
puts "Found #{dist_certs.count} existing distribution certs on App Store Connect."

max_attempts = 5
attempt = 0

while active_cert.nil? && attempt < max_attempts
  attempt += 1
  puts "🚀 Attempt #{attempt}/#{max_attempts} to create paired distribution certificate..."
  begin
    active_cert = Spaceship::ConnectAPI::Certificate.create(
      certificate_type: "DISTRIBUTION",
      csr_content: csr_content
    )
    puts "✅ Created fresh distribution certificate ID: #{active_cert.id}"
  rescue => e
    puts "⚠️ Certificate create response (attempt #{attempt}): #{e.message}"
    dist_certs = Spaceship::ConnectAPI::Certificate.all(filter: { certificateType: "DISTRIBUTION" })
    if dist_certs.any?
      oldest = dist_certs.sort_by { |c| c.expiration_date.to_s }.first
      puts "🗑️ Revoking oldest certificate #{oldest.id} to free up distribution slot..."
      begin
        oldest.delete!
        puts "✅ Successfully revoked cert #{oldest.id}. Waiting 4 seconds for Apple API propagation..."
        sleep(4)
      rescue => del_err
        puts "⚠️ Revoke error: #{del_err.message}"
      end
    else
      sleep(3)
    end
  end
end

if active_cert.nil?
  puts "❌ CRITICAL: Failed to create paired distribution certificate after #{max_attempts} attempts!"
  exit 1
end

if active_cert && File.exist?(dist_key_path)
  File.binwrite(cer_path, Base64.decode64(active_cert.certificate_content))
  system("openssl x509 -inform der -in #{cer_path} -out #{pem_path}")
  system("openssl pkcs12 -export -inkey #{dist_key_path} -in #{pem_path} -out #{p12_path} -passout pass:#{password} -name 'Apple Distribution: Forma'")
  system("security import #{p12_path} -k #{keychain_path} -P #{password} -T /usr/bin/codesign -T /usr/bin/security")
  system("security set-key-partition-list -S apple-tool:,apple:,codesign: -s -k #{password} #{keychain_path}")
  puts "✅ Fresh Apple Distribution certificate imported into keychain!"
end

active_cert_id = active_cert.id
puts "🎯 Active Certificate ID for profiles: #{active_cert_id}"

# Extract exact SHA-1 fingerprint of the imported certificate
cert_sha1 = nil
if File.exist?(cer_path)
  raw_fp = `openssl x509 -inform der -in #{cer_path} -noout -fingerprint -sha1`.to_s
  cert_sha1 = raw_fp.split('=').last.to_s.strip.gsub(':', '')
  puts "🎯 Active Certificate SHA-1 Fingerprint: #{cert_sha1}"
  File.write("/tmp/signing_identity.txt", cert_sha1)
  if ENV['GITHUB_ENV']
    File.open(ENV['GITHUB_ENV'], 'a') { |f| f.puts("SIGNING_IDENTITY=#{cert_sha1}") }
  end
end

# 3. Ensure Bundle IDs and Capabilities
target_bundles = {
  'com.samvel.forma' => 'Forma AppStore',
  'com.samvel.forma.widgets' => 'Forma Widgets AppStore',
  'com.samvel.forma.watchkitapp' => 'Forma Watch AppStore'
}

puts "📋 Ensuring Bundle IDs exist..."
all_bundle_ids = Spaceship::ConnectAPI::BundleId.all

target_bundles.each do |bid_str, prof_name|
  target_bid = all_bundle_ids.find { |b| b.identifier == bid_str }
  if target_bid.nil?
    puts "Creating new Bundle ID: #{bid_str} (#{prof_name})..."
    begin
      target_bid = Spaceship::ConnectAPI::BundleId.create(
        name: prof_name.gsub(' AppStore', ''),
        identifier: bid_str,
        platform: "UNIVERSAL"
      )
      puts "  ✅ Created Bundle ID: #{bid_str}"
    rescue => err
      puts "  ⚠️ Bundle ID creation note: #{err.message}"
    end
  else
    puts "  ✅ Bundle ID exists: #{bid_str}"
  end
  
  if target_bid && bid_str == 'com.samvel.forma'
    ['HEALTHKIT', 'APP_GROUPS'].each do |cap|
      begin
        Spaceship::ConnectAPI::BundleIdCapability.create(
          bundle_id_id: target_bid.id,
          capability_type: cap
        )
        puts "  ✅ Enabled capability #{cap} for #{bid_str}"
      rescue => cap_err
      end
    end
  elsif target_bid && bid_str == 'com.samvel.forma.widgets'
    begin
      Spaceship::ConnectAPI::BundleIdCapability.create(
        bundle_id_id: target_bid.id,
        capability_type: 'APP_GROUPS'
      )
      puts "  ✅ Enabled capability APP_GROUPS for #{bid_str}"
    rescue => cap_err
    end
  end
end

# Refresh Bundle IDs & Profiles
all_bundle_ids = Spaceship::ConnectAPI::BundleId.all
all_profiles = Spaceship::ConnectAPI::Profile.all(includes: "bundleId,certificates")

# 4. Fetch or Recreate Provisioning Profiles explicitly containing active_cert_id
profiles_out_dir = "/tmp/profiles"
FileUtils.mkdir_p(profiles_out_dir)
installed_profiles_dir = File.expand_path("~/Library/MobileDevice/Provisioning Profiles")
FileUtils.mkdir_p(installed_profiles_dir)

puts "📋 Validating and synchronizing Provisioning Profiles with active certificate #{active_cert_id}..."

target_bundles.each do |bid_str, prof_name|
  target_bid = all_bundle_ids.find { |b| b.identifier == bid_str }
  matching_profiles = all_profiles.select do |p|
    p.bundle_id&.identifier == bid_str && 
    (p.profile_type.to_s.include?("APP_STORE") || p.profile_type.to_s.include?("DISTRIBUTION"))
  end
  
  # Check if any matching profile includes active_cert_id
  valid_profile = matching_profiles.find do |p|
    p.certificates&.any? { |c| c.id == active_cert_id }
  end

  # Delete outdated profiles that don't contain active_cert_id
  matching_profiles.each do |p|
    if p != valid_profile
      puts "🗑️ Deleting outdated profile #{p.name} (#{p.id}) because it lacks certificate #{active_cert_id}..."
      begin
        p.delete!
      rescue => del_err
        puts "Note on delete: #{del_err.message}"
      end
    end
  end

  # If no valid profile containing active_cert_id exists, create a fresh one
  if valid_profile.nil? && target_bid && active_cert_id
    puts "Creating fresh App Store profile for #{bid_str} containing certificate #{active_cert_id}..."
    prof_type = bid_str.include?("watch") ? "WATCHOS_APP_STORE" : "IOS_APP_STORE"
    begin
      valid_profile = Spaceship::ConnectAPI::Profile.create(
        name: prof_name,
        profile_type: prof_type,
        bundle_id_id: target_bid.id,
        certificate_ids: [active_cert_id]
      )
      puts "  ✅ Created profile #{prof_name} with cert #{active_cert_id}"
    rescue => err
      puts "  ⚠️ Note on profile create (#{prof_type}): #{err.message}. Retrying as IOS_APP_STORE..."
      begin
        valid_profile = Spaceship::ConnectAPI::Profile.create(
          name: prof_name,
          profile_type: "IOS_APP_STORE",
          bundle_id_id: target_bid.id,
          certificate_ids: [active_cert_id]
        )
      rescue => fallback_err
        puts "  ⚠️ Fallback profile create error: #{fallback_err.message}"
      end
    end
  end

  if valid_profile
    content_b64 = valid_profile.profile_content
    if content_b64
      # Write named profile for direct CI extraction
      named_path = File.join(profiles_out_dir, "#{bid_str}.mobileprovision")
      File.binwrite(named_path, Base64.decode64(content_b64))
      
      # Write UUID profile for Xcode
      uuid_path = File.join(installed_profiles_dir, "#{valid_profile.uuid}.mobileprovision")
      File.binwrite(uuid_path, Base64.decode64(content_b64))
      puts "  ✅ Provisioning Profile verified for #{bid_str} -> #{valid_profile.name} (#{valid_profile.uuid})"
      
      if bid_str == 'com.samvel.forma' && ENV['GITHUB_ENV']
        File.open(ENV['GITHUB_ENV'], 'a') do |f|
          f.puts("PROFILE_NAME=#{valid_profile.name}")
          f.puts("PROFILE_UUID=#{valid_profile.uuid}")
        end
      end
    end
  else
    puts "  ⚠️ Warning: Could not find or create valid profile for #{bid_str}"
  end
end

puts "🎉 Provisioning and Certificate synchronization complete!"

File: scripts/ci_provisioning.py
import os
import sys
import time
import json
import base64
import subprocess

try:
    import requests
    import jwt
except ImportError:
    subprocess.run([sys.executable, "-m", "pip", "install", "--quiet", "pyjwt", "cryptography", "requests", "--break-system-packages"], check=False)
    subprocess.run(["pip3", "install", "--quiet", "pyjwt", "cryptography", "requests"], check=False)
    import requests
    import jwt

def get_auth_token(key_id, issuer_id, key_content):
    if not key_content:
        raise ValueError("App Store Connect API Key is missing")
    
    raw_key = key_content.strip()
    if "BEGIN" not in raw_key:
        try:
            decoded = base64.b64decode(raw_key).decode('utf-8')
            if "BEGIN" in decoded:
                raw_key = decoded
        except Exception:
            pass
            
    if "\\n" in raw_key:
        raw_key = raw_key.replace("\\n", "\n")
        
    now = int(time.time())
    payload = {
        'iss': issuer_id,
        'iat': now - 20,
        'exp': now + 1000,
        'aud': 'appstoreconnect.v1'
    }
    token = jwt.encode(payload, raw_key, algorithm='ES256', headers={'kid': key_id, 'typ': 'JWT'})
    return token

def setup_keychain_with_cert(p12_path, password="buildpassword"):
    keychain_path = "/tmp/build.keychain"
    print(f"🔑 Setting up keychain at {keychain_path}...")
    subprocess.run(["security", "create-keychain", "-p", password, keychain_path], check=False)
    subprocess.run(["security", "set-keychain-settings", "-lut", "21600", keychain_path], check=False)
    subprocess.run(["security", "unlock-keychain", "-p", password, keychain_path], check=False)
    subprocess.run(["security", "import", p12_path, "-k", keychain_path, "-P", password, "-T", "/usr/bin/codesign", "-T", "/usr/bin/security"], check=False)
    subprocess.run(["security", "set-key-partition-list", "-S", "apple-tool:,apple:,codesign:", "-s", "-k", password, keychain_path], check=False)
    
    for url in [
        "https://www.apple.com/certificateauthority/AppleWWDRCAG3.cer",
        "https://www.apple.com/certificateauthority/AppleWWDRCAG6.cer",
        "https://www.apple.com/appleca/AppleIncRootCertificate.cer"
    ]:
        fname = url.split("/")[-1]
        subprocess.run(f"curl -s -f -L '{url}' -o '/tmp/{fname}' && security import '/tmp/{fname}' -k {keychain_path} -T /usr/bin/codesign -T /usr/bin/security", shell=True, check=False)
        
    out = subprocess.check_output(["security", "list-keychains", "-d", "user"]).decode('utf-8')
    existing = [k.strip().replace('"', '') for k in out.split('\n') if k.strip()]
    all_kcs = [keychain_path] + [k for k in existing if k != keychain_path]
    subprocess.run(["security", "list-keychains", "-d", "user", "-s"] + all_kcs, check=False)
    print("✅ Keychain configured successfully.")

def ensure_distribution_certificate(headers, dev_team):
    key_path = "/tmp/dist.key"
    csr_path = "/tmp/dist.csr"
    cer_path = "/tmp/dist.cer"
    pem_path = "/tmp/dist.pem"
    p12_path = "/tmp/dist.p12"
    
    subprocess.run(["openssl", "genrsa", "-out", key_path, "2048"], check=True)
    subprocess.run(["openssl", "req", "-new", "-key", key_path, "-out", csr_path, "-subj", f"/CN=Apple Distribution: Forma/O={dev_team}"], check=True)
    with open(csr_path, 'r') as f:
        csr_content = f.read()

    create_payload = {
        "data": {
            "type": "certificates",
            "attributes": {
                "certificateType": "DISTRIBUTION",
                "csrContent": csr_content
            }
        }
    }
    
    cert_id = None
    max_attempts = 5
    attempt = 0
    res = None
    
    while attempt < max_attempts and cert_id is None:
        attempt += 1
        print(f"🚀 Attempt {attempt}/{max_attempts} to create paired distribution certificate...")
        res = requests.post("https://api.appstoreconnect.apple.com/v1/certificates", json=create_payload, headers=headers)
        if res.status_code in [200, 201]:
            cert_node = res.json().get('data', {})
            cert_id = cert_node.get('id')
            print(f"✅ Created fresh distribution certificate ID: {cert_id}")
            break
        else:
            print(f"⚠️ Certificate create response (attempt {attempt}): {res.status_code} {res.text}")
            list_res = requests.get("https://api.appstoreconnect.apple.com/v1/certificates?filter[certificateType]=DISTRIBUTION&limit=50", headers=headers)
            if list_res.status_code == 200:
                certs = list_res.json().get('data', [])
                certs_sorted = sorted(certs, key=lambda c: c.get('attributes', {}).get('expirationDate', ''))
                if certs_sorted:
                    oldest_id = certs_sorted[0].get('id')
                    print(f"🗑️ Revoking oldest certificate {oldest_id}...")
                    requests.delete(f"https://api.appstoreconnect.apple.com/v1/certificates/{oldest_id}", headers=headers)
                    time.sleep(4)
                else:
                    time.sleep(3)
            else:
                time.sleep(3)

    if not cert_id:
        print("❌ CRITICAL: Failed to create paired distribution certificate!")
        sys.exit(1)

    cert_node = res.json().get('data', {})
    cert_id = cert_node.get('id')
    cert_data = cert_node.get('attributes', {})
    cert_b64 = cert_data.get('certificateContent')
    with open(cer_path, 'wb') as f:
        f.write(base64.b64decode(cert_b64))
    
    subprocess.run(["openssl", "x509", "-inform", "der", "-in", cer_path, "-out", pem_path], check=True)
    subprocess.run(["openssl", "pkcs12", "-export", "-inkey", key_path, "-in", pem_path, "-out", p12_path, "-passout", "pass:buildpassword", "-name", "Apple Distribution: Forma"], check=True)
    setup_keychain_with_cert(p12_path, password="buildpassword")
    
    # Extract SHA-1 Fingerprint
    raw_fp = subprocess.check_output(["openssl", "x509", "-inform", "der", "-in", cer_path, "-noout", "-fingerprint", "-sha1"]).decode('utf-8')
    sha1 = raw_fp.split('=')[-1].strip().replace(':', '')
    print(f"🎯 Distribution SHA-1 Fingerprint: {sha1}")
    with open("/tmp/signing_identity.txt", "w") as f:
        f.write(sha1)
    github_env = os.environ.get('GITHUB_ENV')
    if github_env:
        with open(github_env, 'a') as f:
            f.write(f"SIGNING_IDENTITY={sha1}\n")

    return cert_id

def ensure_bundle_ids_and_profiles(headers, cert_id):
    profiles_out_dir = "/tmp/profiles"
    os.makedirs(profiles_out_dir, exist_ok=True)
    installed_profiles_dir = os.path.expanduser('~/Library/MobileDevice/Provisioning Profiles')
    os.makedirs(installed_profiles_dir, exist_ok=True)

    targets = {
        'com.samvel.forma': 'Forma AppStore',
        'com.samvel.forma.widgets': 'Forma Widgets AppStore',
        'com.samvel.forma.watchkitapp': 'Forma Watch AppStore'
    }

    # 1. Fetch Bundle IDs
    bids_res = requests.get('https://api.appstoreconnect.apple.com/v1/bundleIds?limit=100', headers=headers)
    existing_bids = {}
    if bids_res.status_code == 200:
        for b in bids_res.json().get('data', []):
            existing_bids[b['attributes']['identifier']] = b['id']

    for bid_str, prof_name in targets.items():
        if bid_str not in existing_bids:
            print(f"Creating Bundle ID for {bid_str}...")
            c_res = requests.post('https://api.appstoreconnect.apple.com/v1/bundleIds', json={
                "data": {
                    "type": "bundleIds",
                    "attributes": {
                        "identifier": bid_str,
                        "name": prof_name.replace(" AppStore", ""),
                        "platform": "UNIVERSAL"
                    }
                }
            }, headers=headers)
            if c_res.status_code in [200, 201]:
                existing_bids[bid_str] = c_res.json()['data']['id']

    # 1.1 Enable Capabilities for Bundle IDs
    for bid_str, prof_name in targets.items():
        bid_id = existing_bids.get(bid_str)
        if bid_id:
            caps = []
            if bid_str == 'com.samvel.forma':
                caps = ['HEALTHKIT', 'APP_GROUPS']
            elif bid_str == 'com.samvel.forma.widgets':
                caps = ['APP_GROUPS']
            elif bid_str == 'com.samvel.forma.watchkitapp':
                caps = ['HEALTHKIT']
            for cap in caps:
                requests.post('https://api.appstoreconnect.apple.com/v1/bundleIdCapabilities', json={
                    "data": {
                        "type": "bundleIdCapabilities",
                        "attributes": {"capabilityType": cap},
                        "relationships": {"bundleId": {"data": {"type": "bundleIds", "id": bid_id}}}
                    }
                }, headers=headers)

    # 2. Fetch Profiles with certificates relationship
    prof_res = requests.get('https://api.appstoreconnect.apple.com/v1/profiles?include=bundleId,certificates&limit=100', headers=headers)
    if prof_res.status_code == 200:
        all_profiles = prof_res.json().get('data', [])
        for bid_str, prof_name in targets.items():
            bid_id = existing_bids.get(bid_str)
            if not bid_id:
                continue

            matching_for_bid = [
                p for p in all_profiles 
                if p.get('relationships', {}).get('bundleId', {}).get('data', {}).get('id') == bid_id
                and ('STORE' in p.get('attributes', {}).get('profileType', '') or 'DISTRIBUTION' in p.get('attributes', {}).get('profileType', ''))
            ]

            valid_profile = None
            for p in matching_for_bid:
                cert_refs = p.get('relationships', {}).get('certificates', {}).get('data', [])
                if any(c.get('id') == cert_id for c in cert_refs):
                    valid_profile = p
                else:
                    # Delete outdated profile lacking active cert
                    p_id = p.get('id')
                    print(f"Deleting outdated profile {p.get('attributes', {}).get('name')} ({p_id})...")
                    requests.delete(f"https://api.appstoreconnect.apple.com/v1/profiles/{p_id}", headers=headers)

            if not valid_profile and cert_id:
                print(f"Creating fresh Provisioning Profile for {bid_str} with certificate {cert_id}...")
                p_type = "WATCHOS_APP_STORE" if "watch" in bid_str else "IOS_APP_STORE"
                create_p = requests.post('https://api.appstoreconnect.apple.com/v1/profiles', json={
                    "data": {
                        "type": "profiles",
                        "attributes": {
                            "name": prof_name,
                            "profileType": p_type
                        },
                        "relationships": {
                            "bundleId": {"data": {"type": "bundleIds", "id": bid_id}},
                            "certificates": {"data": [{"type": "certificates", "id": cert_id}]}
                        }
                    }
                }, headers=headers)
                if create_p.status_code in [200, 201]:
                    valid_profile = create_p.json().get('data')

            if valid_profile:
                attrs = valid_profile.get('attributes', {})
                content_b64 = attrs.get('profileContent')
                uuid = attrs.get('uuid')
                name = attrs.get('name')
                if content_b64:
                    with open(os.path.join(profiles_out_dir, f"{bid_str}.mobileprovision"), 'wb') as f:
                        f.write(base64.b64decode(content_b64))
                    with open(os.path.join(installed_profiles_dir, f"{uuid}.mobileprovision"), 'wb') as f:
                        f.write(base64.b64decode(content_b64))
                    print(f"  ✅ Profile ready for {bid_str} -> {name} ({uuid})")
                    if bid_str == 'com.samvel.forma':
                        github_env = os.environ.get('GITHUB_ENV')
                        if github_env:
                            with open(github_env, 'a') as f:
                                f.write(f"PROFILE_NAME={name}\n")
                                f.write(f"PROFILE_UUID={uuid}\n")

def main():
    key_id = os.environ.get('APP_STORE_CONNECT_KEY_ID')
    issuer_id = os.environ.get('APP_STORE_CONNECT_ISSUER_ID')
    dev_team = os.environ.get('DEV_TEAM', 'V7J345DY58')
    api_key_path = os.path.expanduser(f"~/.appstoreconnect/private_keys/AuthKey_{key_id}.p8")
    
    if os.path.exists(api_key_path):
        with open(api_key_path, 'r') as f:
            api_key_content = f.read()
    else:
        api_key_content = os.environ.get('APP_STORE_CONNECT_API_KEY', '')

    print("🔐 Authenticating with App Store Connect API...")
    token = get_auth_token(key_id, issuer_id, api_key_content)
    headers = {
        'Authorization': f'Bearer {token}',
        'Content-Type': 'application/json'
    }

    cert_id = ensure_distribution_certificate(headers, dev_team)
    ensure_bundle_ids_and_profiles(headers, cert_id)

if __name__ == '__main__':
    main()
